### DISCUSSION TITLE
Windows 7 32 bit version needed

### DISCUSSION THREAD

**Discussion by Shakespeare48** [Cuperino/QPrompt-Teleprompter#148] [General]
Hi there, I have to provide a teleprompter for a show and want to use my ancient Windows 7 laptop for the purpose. I want to download a 32 bit version of Qprompt if anyone knows where or how to do that. Thanks!

**Comment by Cuperino** [comment]
H @Shakespeare48,

While it is still theoretically possible to create a 32 bit build of QPrompt I do not have the means to achieve this. All I can do is provide you with some guidance such that figuring out this particular building process doesn't take you as long as it took me to figure out the rest.

Windows 7 32 bit compatible ≈versions of QPrompt will be theoretically possible for as long as QPrompt continues to use Qt framework version 5 undertheath. One of my goals for this year is to complete the transition towards Qt 6. Whan that's dome, builds using Qt 5 may become impossible to create, but only time will tell.

QPrompt needs a minimum of 6 GB of RAM to compile (to build the program). 32 bit systems can afford a maximum of 2 GB of RAM per process on a 4 GB system, meaning 32 bit versions of QPrompt cannot be built on a 32 bit system and instead should be built on a 64 bit system through cross-compilation. It is up to you to figure out how to do this.

It would be be easier to build a 32 bit version of QPrompt from the 32 bit system, but the only way to achieve this, is by partially dropping Chinese support, by modifying QPrompt's code itself and removing the embedded Chinese font. Doing so will reduce the amount of RAM needed to build QPrompt into something low enough to build from 32 bit hardware.

If you intend to distribute this new build in a usable form, you'll also have to figure out how to create an installer. Our current installer creation methodology is 64 bit only.

To create a QPrompt build for 32 bit Windows you'll need to:

1. Install CMake and Git
2. Install Microsoft VisualStudio 2019 or later version until early 2022, with: `Desktop Development with C++`, `C++ ATL`, `Windows SDK`
3. Install Qt 5 with 32 bit support (for a perfect build, you must compile Qt from source with the most recent changes from KDE's Qt Patch Collection applied to it.)
4. In Qt Creator, setup a build environment which can cross compile to 32 bit
5. Git clone QPrompt repository to your system
6. Open a Command Prompt with Administrator Privileges and with it do most of the remaining steps.
7. Build and cross compile each of the following libraries to 32 bit architecture, and install them (point CMake towards to) into the aforementioned build environment (in the following order):
- The Qt framework (if you haven't already done so)
- KDE's extra-cmake-modules
- KDE's kcoreaddons
- KDE's ki18n
- KDE's Kirigami
- KDE's KCrash and all of it's dependencies (KCrash is an optional dependency and itself has more dependencies than Qt itself)
- Skycoder42's QHotkeys library, provided on GitHub
8. Proceed to build QPrompt. You should use the same commands you've used to build all the previous libraries, which should look similar to this:
```
mkdir build
cmake [ADDITIONAL_INSTRUCTIONS_GO_HERE] -B build/ .
cmake --build build/
```
Once QPrompt has built successfully, you will find its binaries somewhere in the build sub-folder you've created with mkdir. If the build works, you're done. If it doesn't, attempt to launch from the command line to figure out what went wrong and fix it. There's no guarantee a 32 bit build of QPrompt for Windows will even work. 32 bit builds are possible and have been tested to work under 32 bit Linux, but none have ever been released due to our Linux installer package creation methods also being incompatible with 32 bit.

Should you desire to continue, I wish you best of luck and perseverance.
